parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/HadoopStreams.java

```diff
@@ -37,6 +41,39 @@ public class HadoopStreams {
 
   private static final Logger LOG = LoggerFactory.getLogger(HadoopStreams.class);
 
+  private static final Class<?> byteBufferReadableClass = getReadableClass();
+  static final Constructor<SeekableInputStream> h2SeekableConstructor = getH2SeekableConstructor();
```

Comment:
Rather than using two static methods, you can use `DynConstructors` instead to make this one expression and reduce error handling code:

```java
private static final DynConstructors.Ctor<SeekableInputStream> h2streamCtor =
    DynConstructors.Builder(SeekableInputStream.class)
    .impl("org.apache.parquet.hadoop.util.H2SeekableInputStream", FSDataInputStream.class)
    .orNull()
    .build()

...
if (h2streamCtor != null) {
  return h2streamCtor.newInstance(stream);
}
```

Reply:
Looks like you'd need to add the `orNull` handling. I don't see it here: https://github.com/apache/parquet-mr/blob/master/parquet-common/src/main/java/org/apache/parquet/util/DynConstructors.java

Reply:
Huh, I guess this was how it was before? Nevermind on the refactoring then.

Reply:
Yes, I copied most from the old code to avoid refactoring. I think we can greatly simplify it because it was still taking Hadoop1 into account. We still have to check if the wrapped stream is ByteBufferReadable: https://github.com/apache/hadoop/blob/release-2.4.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FSDataInputStream.java#L142-L148

The `hasCapabilities does the same but in a more elegant way.